feat(core): Add routing solution

maciejopalinski opened the pull request:
Done with @T0ffik

maciejopalinski committed:
fix: Update `.prettierrc` formatting rules

maciejopalinski committed:
refactor: Update `index.html`

maciejopalinski committed:
feat: Add `typescript` package

Parcel did this automatically

maciejopalinski committed:
feat: Add `onclick` event listener to `index.js` to handle routing

maciejopalinski committed:
feat: Add example components

maciejopalinski committed:
feat: Add example pages

maciejopalinski committed:
feat: Add `Router` page

maciejopalinski committed:
fix: Fix `render()` method to work with routing solution

maciejopalinski committed:
feat: Render `Router` from `App`

maciejopalinski committed:
format: Format files using `Prettier`

maciejopalinski committed:
refactor: Remove empty files

T0ffik commented:
I see you changed most of js files to ts and I'm worried if the project won't be in mess when other coworkers will type in js?

maciejopalinski commented:
> I see you changed most of js files to ts and I'm worried if the project won't be in mess when other coworkers will type in js?

I think that we should just write in TypeScript. Only thing that I changed is the file extension and that one type definition of arguments that you need to pass into a component. I did that for a reason. When I changed something, plain JS was not catching errors while building the project so I had no idea that the errors were even there. That type definition is not even required.
Just *.ts file extension provides you with compile time errors. And that is much better. It speeds up the development process.

The worst thing about plain JS is that even my IDE does NOT report those errors for example if I pass a string in place of which should be an object.

T0ffik commented:
> > I see you changed most of js files to ts and I'm worried if the project won't be in mess when other coworkers will type in js?
> 
> I think that we should just write in TypeScript. Only thing that I changed is the file extension and that one type definition of arguments that you need to pass into a component. I did that for a reason. When I changed something, plain JS was not catching errors while building the project so I had no idea that the errors were even there. That type definition is not even required. Just *.ts file extension provides you with compile time errors. And that is much better. It speeds up the development process.
> 
> The worst thing about plain JS is that even my IDE does NOT report those errors for example if I pass a string in place of which should be an object.

To be honest I don't TypeScript, so I don't know differences between this and JavaScript, and probably no one of rest too. We gonna learn TypeScript in 4th module/project so I don't think that's the right time to use it. I think we should stay with JavaScript.

maciejopalinski commented:
> > > I see you changed most of js files to ts and I'm worried if the project won't be in mess when other coworkers will type in js?
> > 
> > 
> > I think that we should just write in TypeScript. Only thing that I changed is the file extension and that one type definition of arguments that you need to pass into a component. I did that for a reason. When I changed something, plain JS was not catching errors while building the project so I had no idea that the errors were even there. That type definition is not even required. Just *.ts file extension provides you with compile time errors. And that is much better. It speeds up the development process.
> > The worst thing about plain JS is that even my IDE does NOT report those errors for example if I pass a string in place of which should be an object.
> 
> To be honest I don't TypeScript, so I don't know differences between this and JavaScript, and probably no one of rest too. We gonna learn TypeScript in 4th module/project so I don't think that's the right time to use it. I think we should stay witch JavaScript.

Look there is no difference for you except file extension. Everything else works the same you just get your errors reported on compilation. With plain JS errors only display at runtime in browser when the code is run. It is harder to debug code with plain JS. You don't need to learn any TypeScript for now.

Veryyapeee commented:
Generally,

1. You always have to be able to write code in vanilla js, no matter how painfull it is, in real job no one actually would care if TypeScript creates better development experience, JS developers are cheaper and not every project on the world uses TypeScript

2. If any member of the team disagree with using TypeScript, because it's confusing for them, or anything, we should give up on this, and put it into discussion,
Member of the team don't actually have to learn about it and provide special arguments, it's a feature, not a requirement, and if anyone wants to focus on plain JS, we should respect that

There will be special module on camp about TS

Veryyapeee commented on src/index.js:
```diff
@@ -1,7 +1,16 @@
 
-const BASE_API_URL = process.env.API_BASE_URL || 'https://google.com';
+window.addEventListener('click', (e) => {
+    if (e.target instanceof HTMLAnchorElement) {
```
Can't se an option to pass a parameter to the route, like asset id which we will need

maciejopalinski committed:
refactor: Remove TypeScript from project

maciejopalinski committed:
fix: Fix back arrow behavior in browser

maciejopalinski committed:
feat: Add solution for parameters in routes

Veryyapeee commented on package.json:
```diff
@@ -43,7 +43,8 @@
     "rimraf": "^3.0.2",
-    "sass": "^1.44.0"
+    "sass": "^1.44.0",
+    "typescript": "^4.5.2"
```
remove typescript

Veryyapeee commented on src/pages/Router.js:
```diff
@@ -0,0 +1,39 @@
+
+    const options = { renderOn: '#app-main' };
+
+    if (endpoints[0] == '') {
```
Just
if(!endpoint[0])
Also it could be a switch case instead of if

Veryyapeee commented on src/pages/Router.js:
```diff
@@ -0,0 +1,39 @@
+
+    if (endpoints[0] == '') {
+        HomePage(options);
+    } else if (endpoints[0] == 'search') {
```
`===`

Veryyapeee commented:
Generally looks great, good jub guys 8)

Some minor comments in code

Also I'd like to request one more thing, 
Generally operating on plain strings in code is not a bad practice, it's not maintainable
We should create a routes.js file, and there declare an object with routs that we will be using
Something like: 
`export const routes = {
home: 'home',
search: 'search' 
}`
and so on

Veryyapeee commented on src/pages/Router.js:
```diff
@@ -0,0 +1,39 @@
+export const Router = ({ renderOn }) => {
+    render({ html: template, on: renderOn });
+
+    Navbar({ renderOn: '#app-navbar' });
```
This shouldn't be here

We have task for preparing the <BaseScreenView />

Which will have a Footer and Navbar, and there's will be a div rendered in which we will injecting pages

Would be great if you could seperate these two logics, @wikbro7650 is handling the BaseView task

maciejopalinski replied:
I don't really understand what's wrong here. I just created a dummy example component. What do you want me to do? I don't think that I should remove those examples because it won't be clear how to work with Router component.

Veryyapeee replied:
1. If anything is temporaty there should be comments with explenation what is what and why
And TODO comments about code which have to be removed in the future

2. My idea is to
- router shouldn't contain any hardcoded HTML or anything like that
-router should be mounted in BaseView
-Router will accept a parametr of array of objects, ex. each object will have a route name which is assigned to, and the component itselft
Based on that we can just pass an array of object and dynamically render each route, without hardcoding like it is now

Something like, you are passing an array, object, whatever you find best, ex. On object

Screens: {
'/home':{
component: HomeScreen
on: '#something'
}
}

Then in router you can just do

Screens[currentRoute].component(...)

And boom, we have dynamic logic, and Route is independent function without any html in it

About the strukture, you can go with object, array, anything that suites you

If you need more explenation or help we can go on a call tomorrow

maciejopalinski replied:
Ok this makes sense. Will do!

Veryyapeee reviewed (changes requested):
As in the comments :P

maciejopalinski committed:
fix: Fix routing solution to use Regex

maciejopalinski committed:
fix: Fix `SearchPage` component to use new routing solution

maciejopalinski committed:
fix: Fix `App` component to use new routing solution

maciejopalinski committed:
refactor: Remove examples from code

As I was told to do.

maciejopalinski committed:
fix(deps): Remove `typescript` package

maciejopalinski committed:
Squashed commit of the following:

commit e27d79039c18b867e987475d5b690e0af7b3beaf
Author: Maciej Opaliński <poprostumieciek@gmail.com>
Date:   Mon Dec 6 22:24:24 2021 +0100

    feat(api): Add Movie Database API integration (#7)

    * refactor(core): Rename `.env` to `.env.example`

    * refactor(core): Add `.env` to `.gitignore`

    * fix: Change `RAPID_API_KEY` to `API_SECRET_KEY` in `.env.example`

    * fix: Fix `API_BASE_URL` variable to use new API

    * feat: Add `USE_API` env variable

    * feat!: Update `.prettierrc` formatting rules

    * fix: Cleanup and format some files

    * feat: Add `utils/Utils.js`

    * feat: Add `CacheManager.js`

    * feat: Add `ApiService.js`

    * feat!: Add cache entries

    * feat: Add example `ApiService` usage to `App.js`

    !!! This is temporary just to check if ApiService works. !!!
    You will not use this class directly!
    There will be another class ApiProvider that you will work with.

    * fix(deps): Fix `package-lock.json`

    NPM did this automatically.

    * Revert "fix(deps): Fix `package-lock.json`"

    This reverts commit b65ccf2a50105f2410e4d39a72fdc849928f36ae.

    * fix: Fix `USE_API` variable comparison in `ApiService`

    * fix: Update imports in `ApiService`

    * feat: Add `ApiProvider` class

    * feat: Update examples in `App` to use `AppProvider`

    * feat: Add cache entries

    * feat: Add `reportNewCacheEntry` method to `ApiService`

    * fix: Update example in `App`

    * feat: Add new cache entries

    * feat!: Add `Project Setup` chapter to `README.md`

    * feat!(deps): Add `typescript` package

    * chore: Fix formatting according to `.prettierrc`

    * feat: Add TypeScript interfaces for API requests

    * refactor: Add strict typing to `ApiProvider`

    * refactor: Add strict typing to `ApiService`

    * refactor: Add strict typing to `CacheManager`

    * refactor: Add strict typing to methods in `utils/`

    * feat: Add new cache entries

    * fix: Change `API_BASE_URL`

    * feat: Add TypeScript interfaces for API requests

    * feat: Remove old cache entries and add new ones

    * refactor: Refactor `Api*` classes to use new API

    * fix: Update example usage in `App` component

    * fix: Remove unused `utils/` directory

    * fix: Fix `API_BASE_URL` env variable

    * feat: Add TypeScript interfaces for API requests

    * feat: Add auto file downloading to `CacheManager`

    * fix: Fix `ApiService`

    * fix: Fix cache entries

    * fix: Fix example API usage in `App` component

    * fix!: Update `README.md`

    * fix: Fix variable names and access specifiers

    * refactor: Remove TypeScript from project

    * fix: Remove `typescript` package

    * fix: Remove usage example from `App.js`

maciejopalinski committed:
Merge branch 'main' into feature/routing

Veryyapeee commented on src/shared/dom.js:
```diff
@@ -1,13 +1,19 @@
+        return element;
+    } else {
+        placeholder.replaceChildren(element);
+        return element;
```
`if() {}....
else{}...
return element
`

Veryyapeee commented on src/pages/Router.js:
```diff
@@ -0,0 +1,18 @@
+import { render } from '../shared/dom';
+
+export const Router = ({ routes, on }) => {
```
Just two more things and you are good to go 8)

create routes.js file with empty object like so with one example
`export const routes = {
home: 'home'
}`

In comment define how the routes object should look like (above the Router component ifselt)

maciejopalinski committed:
fix: Remove unnecessary imports

maciejopalinski committed:
fix: Simplify control flow in `render()`

maciejopalinski committed:
refactor: Simplify and move `Router` to `/route` dir

maciejopalinski committed:
feat: Add example usage in `App.js`

Veryyapeee reviewed (approved):
LGTM!!!11!1!!! 8)
